fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ScalarApply2AnalyticRule.java

```diff
@@ -0,0 +1,869 @@
+                return false;
+            }
+            {
+                Iterator<ScalarOperator> outerIt = outerConjuncts.iterator();
+                while (outerIt.hasNext()) {
+                    ScalarOperator outerConjunct = outerIt.next();
+                    Iterator<ScalarOperator> subIt = subConjuncts.iterator();
+                    boolean found = false;
+                    while (subIt.hasNext()) {
+                        ScalarOperator subConjunct = subIt.next();
+                        if (PredicateComparator.isIdentical(outerConjunct, subConjunct,
+                                context.getColumnRefFactory())) {
+                            found = true;
+                            outerIt.remove();
+                            subIt.remove();
+                            break;
+                        }
+                    }
+                    if (!found) {
+                        return false;
+                    }
+                }
+            }
+            Preconditions.checkState(outerConjuncts.isEmpty());
+            Preconditions.checkState(subConjuncts.isEmpty());
+
+            return true;
+        }
+
+        private void createWindowOp() {
+            Map<ColumnRefOperator, CallOperator> windows = Maps.newHashMap();
+
+            Map<ColumnRefOperator, CallOperator> aggregations = subAggOp.getAggregations();
+
+            for (Map.Entry<ColumnRefOperator, CallOperator> entry : aggregations.entrySet()) {
+                ColumnRefOperator columnRefOperator = entry.getKey();
+                CallOperator callOp = entry.getValue();
+
+                CallOperator newCallOp = cloneCallOperator(callOp);
+                ColumnRefOperator newColumnRefOperator = context.getColumnRefFactory()
+                        .create(newCallOp, newCallOp.getType(), newCallOp.isNullable());
+                windows.put(newColumnRefOperator, newCallOp);
+                peerColumnRefMap.put(columnRefOperator, newColumnRefOperator);
+            }
+
+            List<ScalarOperator> partitionByColumns = Lists.newArrayList();
+            for (ScalarOperator outerCorrelatedConjunct : outerCorrelatedConjuncts) {
+                Preconditions.checkState(outerCorrelatedConjunct instanceof BinaryPredicateOperator);
+                // We can take either first or second child of outerCorrelatedConjunct as the partition by expr
+                partitionByColumns.add(outerCorrelatedConjunct.getChild(0));
+            }
+
+            windowOp = new LogicalWindowOperator.Builder()
+                    .setWindowCall(windows)
+                    .setPartitionExpressions(partitionByColumns)
+                    .setEnforceSortColumns(
+                            partitionByColumns.stream()
+                                    .map(ScalarOperator::<ColumnRefOperator>cast)
+                                    .map(columnRefOperator -> new Ordering(columnRefOperator, true, true))
```

Comment:
why the isAsc and isNullFirst are both true?

Reply:
It does not matter, any selection can work out.